### PR TITLE
Expose gradients w.r.t. input & weight for conv1d, conv2d, conv3d in Python

### PR DESCRIPTION
**Pull request by vedanuj**
This PR addresses issue #5024 and exposes in python gradients with respect to input and weight for `conv2d` and `conv3d`. 

### Gradients with respect to input of convolution (`F.grad.conv2d_input` , `F.grad.conv3d_input`) : 

For this `conv_transpose2d`/`conv_transpose3d` are used which are essentially the gradients of `conv2d`/`conv3d` w.r.t. input. The output_padding required for `conv_transpose2d`/`conv_transpose3d` are implicitly calculated from shape of input gradient, shape of output gradient, kernel size, stride and padding.

### Gradients with respect to weights (`F.grad.conv2d_weight` , `F.grad.conv3d_weight`): 

For gradients with respect to the weight we use the  `conv2d`/`conv3d` but with strides and dilations swapped. We reshape the output gradient and input in such a way that the output gradient can be treated as the weight of the forward convolution.

Please review @ezyang

### CODE REVIEW

File: torch/nn/modules/conv.py [hunk 1]
```diff
@@ -636,6 +636,102 @@ def forward(self, input, output_size=None):
             output_padding, self.groups, self.dilation)
 
 
+class Conv2dBackward(_ConvTransposeMixin, _ConvNd):
```
Comment (apaszke) [inline torch/nn/modules/conv.py]: I don't think that adding this module is a good idea. First, as noticed in this docstring, it's pretty much just conv transpose. Secondly, once we have `Conv2dBackward`, why shouldn't we also add `Conv3dBackward`, `LinearBackward`, `ReLUBackward`, etc...

What's the use case for this? Maybe we should just have an interface for accessing the gradients we implement in autograd, but as a functional interface that doesn't need the first stage to happen and capture the relevant elements?

Comment (vedanuj) [reply torch/nn/modules/conv.py]: @apaszke The issue #5024 discusses some use cases where they require alternate forms of backprop through convolution. This API is functionally similar to what TensorFlow's https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_backprop_input provides. Tensorflow also uses the same underlying API for both conv2d_backprop_input and [conv2d_transpose](https://github.com/tensorflow/tensorflow/blob/54ab9d690a63ee11813f091d9dd9c2cd3c4cce1d/tensorflow/python/ops/nn_ops.py#L1183). 

Do you suggest we should make a generic interface that can call into the backward pass for any operator function?

Comment (ezyang) [reply torch/nn/modules/conv.py]: So, the real utility of exposing backwards convolution is not the backwards on input (which is pretty simple) but the backwards on weight, which is also itself a convolution but with dilation and stride swapped. Yes, the user can do the math themselves, but why not just make their life easier.

As for the slippery slope argument, I think the answer is the same as all the other ops; we add them if people ask for them, and don't if they don't. A functional interface to autograd sounds like an awfully convoluted way to do this. When we have the symbolic differentiator things are probably less bad.

@vedanuj We're probably going to need all of the other variants, but maybe we should first convince @apaszke this is a good idea :)

Comment (apaszke) [reply torch/nn/modules/conv.py]: I agree that having *some* way of supporting people who need this would be nice. I'm just saying that making the nn API 2x larger doesn't seem like a great solution, especially considering that it doesn't solve the original problem mentioned in that forum thread:

> I want to use a coustom forward function with the standard convolutional backward function.
> Maybe I should use the standard function when I need a backpropagation. Thank you.

Implementing a backward of an autograd function using a module seems like a bad idea to me. Adding it to functional would be better, but I still don't like the fact that it's another function we have to maintain manually, when we're auto-generating them for internal purposes.

Symbolic differentiation is never meant to be exposed to users, and is designed for internal use in the JIT only. We don't even plan to expand its scope to cover all of autograd in the core library.

Comment (ezyang) [reply torch/nn/modules/conv.py]: > Implementing a backward of an autograd function using a module seems like a bad idea to me. Adding it to functional would be better, but 

Having just the functional seems fine to me.

> I still don't like the fact that it's another function we have to maintain manually, when we're auto-generating them for internal purposes.

But we're not auto-generating them! We have explicitly implemented kernels for each of the backwards. We're just giving aliases to them to help users out.

Comment (apaszke) [reply torch/nn/modules/conv.py]: Fine, let's do the functional version only, like TF does it.

**Comment by soumith** [comment]
So, the reservation from Adam is valid, but not in the way he expressed it.
What we want is to not increase the API surface to accommodate each of these "backward" calls.
Imagine having:

```
F.conv2d
F.conv2d_backward_input
F.conv2d_backward_weight
F.conv2d_backward_bias
```

I think the right design would be a separate lookup table that branches one step further. For example:

```
F.grad.conv2d_input
F.grad.conv2d_weight
F.grad.conv2d_bias
```

This way, all this extra API is encapsulated in the `grad` tree.

**Comment by ezyang** [comment]
`grad` is just a module, is that right? SGTM. @vedanuj, does this make sense?

**Comment by soumith** [comment]
@vedanuj correct

**Comment by apaszke** [comment]
I thought about the namespace, and it's not a significantly better solution (the complexity remains the same), but I'm ok with it. One extra thing that we might want to think of before we start adding more and more functions there is that sometimes the derivative computations for a single op can share a lot of intermediates, and so we might need to redesign the API (to avoid having a function per computed gradient)

File: torch/nn/gradients.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+import torch
+from .modules.utils import _pair, _triple
+
+
+def _output_padding(input, output_size, stride, padding, kernel_size):
+    output_size = list(output_size)
+    k = input.dim() - 2
+    print(k)
+    print(output_size)
```
Comment (ezyang) [inline torch/nn/gradients.py]: X these

File: torch/nn/functional.py [hunk 3]
```diff
@@ -191,6 +192,10 @@ def conv_tbc(input, weight, bias, pad=0):
     return input.conv_tbc(weight, bias, pad)
 
 
+# Interface for the gradients of various operators
+grad = G
```
Comment (apaszke) [inline torch/nn/functional.py]: If a user wants a quick access to a module by aliasing it with such a short name then fine, but we shouldn't do it for them. Let's remove this line

File: torch/nn/gradients.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        padding (int or tuple, optional): Zero-padding added to both sides of the input. Default: 0
+        dilation (int or tuple, optional): Spacing between kernel elements. Default: 1
+        groups (int, optional): Number of blocked connections from input channels to output channels. Default: 1
+        bias: optional bias tensor (out_channels). Default: None
+
+    Examples::
+
+        >>> out_backprop = autograd.Variable(torch.randn(1,1,3,2))
+        >>> weight = autograd.Variable(torch.randn(1,1,1,2))
+        >>> F.grad.conv2d_input(out_backprop, weight, stride=1, input_size=(1,1,3,3))
```
Comment (ezyang) [inline torch/nn/gradients.py]: A helpful thing to show is how `F.grad.conv2d_input` is related to a sequence of calls to `F.conv2d` and then `torch.autograd.grad()`. These would be good tests as well (which we need to have on the PR.)

File: torch/nn/gradients.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+def _output_padding(input, output_size, stride, padding, kernel_size):
+    output_size = list(output_size)
+    k = input.dim() - 2
+    print(k)
+    print(output_size)
+    if len(output_size) == k + 2:
+        output_size = output_size[-k:]
+    if len(output_size) != k:
+        raise ValueError("output_size must have {} or {} elements (got {})"
+                         .format(k, k + 2, len(output_size)))
```
Comment (ezyang) [inline torch/nn/gradients.py]: Why this degree of freedom?

File: torch/nn/gradients.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            raise ValueError(
+                ("requested an output size of {}, but valid sizes range "
+                 "from {} to {} (for an input of {})").format(
+                     output_size, min_sizes, max_sizes,
+                     input.size()[2:]))
+
+    return tuple([output_size[d] - min_sizes[d] for d in range(k)])
+
+
+def conv2d_input(in_channels, out_channels, out_backprop, weight, input_size,
```
Comment (ezyang) [inline torch/nn/gradients.py]: I like the other convention better, where you have `input_size, weight`. The idea is that the relative ordering of input and weight is always the same.

Comment (ezyang) [reply torch/nn/gradients.py]: This would also be consistent with TF's argument ordering https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/nn/conv2d_backprop_input

File: torch/nn/gradients.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        >>> out_backprop = torch.autograd.Variable(torch.randn(1,1,3,2))
+        >>> F.grad.conv2d_weight(1,1, input, out_backprop)
+    """
+    stride = _pair(stride)
+    padding = _pair(padding)
+    dilation = _pair(dilation)
+    dims = out_backprop.shape
+    min_batch = input.shape[0]
+
+    out_backprop_ = out_backprop.repeat(1, in_channels // groups, 1, 1)
```
Comment (ezyang) [inline torch/nn/gradients.py]: The code here looks more complicated than I thought it should be. What's going on here?

**Comment by ezyang** [comment]
Needs tests.

File: torch/nn/gradients.py (hunk 7 above)
Comment (vedanuj) [reply torch/nn/gradients.py]: For calculating the gradient with respect to weight, the `out_backprop` is the weight for the `conv2d` and the weight gradient is the output.  The `out_backprop` is expanded and reshaped so that the forward convolution results in the (weight gradient) having the same shape as the `weight`. The `out_backprop` is expanded by the number of input channels divided by the number of groups (`in_channels // groups`). Then the `input` is reshaped so that it has channels and batches merged and the `conv2d` forward pass with groups = `in_channels * min_batch`. 

The result is reshaped with dimension 0 as the number of batches and the tensor is summed along that dimension. After that, the result(`grad_w`) is reshaped and narrowed to fit the `weight_size`.

**Comment by vedanuj** [comment]
The last commit added tests for all the grad functions and addresses other comments.

### Tests: 

`python test/test_nn.py TestNN.test_grad_conv2d_input`

`python test/test_nn.py TestNN.test_grad_conv2d_weight`

`python test/test_nn.py TestNN.test_grad_conv3d_input`

`python test/test_nn.py TestNN.test_grad_conv3d_weight`


@ezyang  @apaszke Please re-review it.

File: test/test_nn.py [hunk 8]
```diff
@@ -4305,6 +4305,93 @@ def test_conv_double_backward_cuda(self):
                                 "\ninp_size: " + str(inp_size) +
                                 "\ndilation: " + str(dilation))
 
+    def test_grad_conv2d_weight(self):
```
Comment (ezyang) [inline test/test_nn.py]: Is there a way to not duplicate the code here? For example, you are also going to need 1d as well

**Comment by vedanuj** [comment]
The commit 23a6b3a adds conv1d gradient w.r.t. input(`F.grad.conv1d_input`) and weight(`F.grad.conv1d_weight`) and refactors tests for all the conv grad functions.

**Comment by vedanuj** [comment]
Please review @ezyang @apaszke

File: torch/nn/gradients.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
+"""Gradient interface"""
+
+import torch
+from .modules.utils import _single, _pair, _triple
+
+
+def input_grad_padding(out_backprop, input_size, stride, padding, kernel_size):
```
Comment (apaszke) [inline torch/nn/gradients.py]: This should probably be private (prefix with an underscore)

File: test/test_nn.py [hunk 10]
```diff
@@ -4305,6 +4305,51 @@ def test_conv_double_backward_cuda(self):
...
+                    product([1, 2], [1, 2], [0, 1, 2], [2], [3], [1]):
+
+                input_shape = [batch, chan_in]
+                weight_shape = [chan_out, chan_in]
+                for _ in range(dim):
+                    input_shape.append(inp_size)
+                    weight_shape.append(kern)
+
+                input = torch.autograd.Variable(torch.randn(input_shape), requires_grad=True)
+                weight = torch.autograd.Variable(torch.randn(weight_shape), requires_grad=True)
```
Comment (apaszke) [inline test/test_nn.py]: `Variable` is deprecated. You can pass `requires_grad=True` directly to `.randn`

File: torch/nn/gradients.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+
+
+def conv1d_input(input_size, weight, out_backprop, stride=1, padding=0, dilation=1, groups=1, bias=None):
+    r"""
+    Computes the gradient of conv1d with respect to the input of the convolution.
+    This is same as the 1D transposed convolution operator under the hood but requires
+    the shape of the gradient w.r.t. input to be specified explicitly.
+
+    Args:
+        input_size : Shape of the input gradient tensor
```
Comment (apaszke) [inline torch/nn/gradients.py]: Wouldn't it be possible to require something simpler than the full input shape and infer some parameters? E.g. you can certainly do this with #channels, but the sizes might be ambiguous, right?

File: torch/nn/gradients.py [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+
+
+def conv3d_weight(input, weight_size, out_backprop, stride=1, padding=0, dilation=1, groups=1, bias=None):
+    r"""
+    Computes the gradient of conv3d with respect to the weight of the convolution.
+
+    Args:
+        input: input tensor of shape (minibatch x in_channels x iT x iH x iW)
+        weight_size : Shape of the weight gradient tensor
+        out_backprop : output gradient tensor (minibatch x out_channels x oT x oH x oW)
```
Comment (apaszke) [inline torch/nn/gradients.py]: Please call this `grad_output`. That's the convention we have everywhere else

**Review by apaszke — COMMENTED** [review]
LGTM, but I have some minor comments

File: torch/nn/gradients.py (hunk 11 above)
Comment (vedanuj) [reply torch/nn/gradients.py]: Yes we can do that. We can calculate the batch and channels of input from weight and output shape.

**Comment by vedanuj** [comment]
New commits changes some names of parameters and removes use of deprecated `Variable`. 

@apaszke TF requires to input the full shape for input size and weight size. Should we keep it same as TF or change input size and weight size to only specify the dimensions like height, width, depth?

**Comment by apaszke** [comment]
No, if we're going to accept the shape then it should be the full shape. Let's leave it as is.
